profile.d/00-pengwin.sh

```diff
@@ -1,5 +1,22 @@
-# enable external x display
-export DISPLAY=:0
+# check whether it is WSL1 for WSL2 by using gcc version kernel built with
+# check whether we have wsl.exe in path
+if ( which wsl.exe >/dev/null ) && ( $(which wsl.exe) -l -v >/dev/null ) && [ $(wsl.exe -l -v 2>&1 |sed -e "s|\r||g" -e "s|\x00||g" | grep ${WSL_DISTRO_NAME} | awk '{print $3}') -eq 2 ]; then
+  # enable external x display for WSL 2
+
+  ipconfig_exec=$(wslpath "C:\\Windows\\System32\\ipconfig.exe")
+  if ( which ipconfig.exe >/dev/null ); then
+    ipconfig_exec=$(which ipconfig.exe)
+  fi
+
+  wsl2_d_tmp="$(eval "$ipconfig_exec" | grep -n "Default Gateway.*: [0-9]" | cut -d : -f 1)"
```

Comment:
"Default Gateway" could be language-specific.
![image](https://user-images.githubusercontent.com/6897531/70299649-d4769d80-1838-11ea-9f48-e19fc51f5398.png)

Reply:
I was wrong, it become English if output is redirected.
![image](https://user-images.githubusercontent.com/6897531/70322060-66e46480-186c-11ea-8deb-89df5b18c82c.png)

Reply:
Could you please replace your profile.d/00-pengwin.sh by the one in this PR, and make some tests? Backup your first. 

Thank you

Reply:
Sorry, I can't run actual test as I don't have WSL2.
However, I found that `Default Gateway.*: [0-9]` doesn't match "Default Gateway" line in my `ipconfig.exe` output because my default gateway address starts with `fe80::` which is hexadecimal number.

Reply:
I tried in pengwin with WSL2, it doesn't work.  I got this error line:

> -bash: [: Running: integer expression expected

Reply:
@eliranwong Could you please test again? I made some modifications, also added debug mode. If it doesn't work could you please send me the output of your ipconfig.exe? Regards

Reply:
@crramirez 

I tried with your new fix.  It doesn't work.  Below is the output in terminal.  I sent you my output of ipconfig to your gmail, I think it is better to keep my ip from very public:

++ which wsl.exe
+++ which wsl.exe
++ /mnt/c/WINDOWS/system32/wsl.exe -l -v
+++ wsl.exe -l -v
+++ sed -e 's|\r||g' -e 's|\x00||g'
+++ grep WLinux
+++ awk '{print $3}'
++ '[' Running -eq 2 ']'
-bash: [: Running: integer expression expected
++ export DISPLAY=:0
++ DISPLAY=:0
++ _pyenv_virtualenv_hook
++ local ret=0
++ '[' -n '' ']'
+++ pyenv sh-activate --quiet
+++ local command
+++ command=sh-activate
+++ '[' 2 -gt 0 ']'
+++ shift
+++ case "$command" in
+++ command pyenv sh-activate --quiet
+++ pyenv sh-activate --quiet
+++ true
++ eval false
+++ false
++ true
++ return 0
++ __vte_prompt_command
++ local 'pwd=~'
++ '[' /home/eliran '!=' /home/eliran ']'
++ pwd='~'
+++ __vte_osc7
++++ __vte_urlencode /home/eliran
++++ LC_ALL=C
++++ str=/home/eliran
++++ '[' -n /home/eliran ']'
++++ safe=/home/eliran
++++ printf %s /home/eliran
++++ str=
++++ '[' -n '' ']'
++++ '[' -n '' ']'
+++ printf '\033]7;file://%s%s\033\' everlasting /home/eliran
++ printf '\033]0;%s@%s:%s\033\%s' eliran everlasting '~' ''

Reply:
@crramirez , fyi, I am using surface go (LTE version)

Reply:
@crramirez I think here may be where the issue is, I am not sure.

when I entered:
> wsl.exe -l -v

I got this output:
  NAME      STATE           VERSION
* WLinux    Running         2
  WLE       Stopped         1

HOWEVER, when I entered:
> wsl.exe -l -v | grep $WSL_DISTRO_NAME
there is NO output at all

I wonder why, so I entered:
> wsl.exe -l -v > test.txt
> nano test.txt

Below is the output:
 ^@ ^@N^@A^@M^@E^@ ^@ ^@ ^@ ^@ ^@ ^@S^@T^@A^@T^@E^@ ^@ ^@ ^@ ^@ ^@ ^@ ^@ ^@ ^@ ^@ ^@V^@E^@R^@S^@I^@O^@N^@
^@
^@*^@ ^@W^@L^@i^@n^@u^@x^@ ^@ ^@ ^@ ^@R^@u^@n^@n^@i^@n^@g^@ ^@ ^@ ^@ ^@ ^@ ^@ ^@ ^@ ^@2^@
^@
^@ ^@ ^@W^@L^@E^@ ^@ ^@ ^@ ^@ ^@ ^@ ^@S^@t^@o^@p^@p^@e^@d^@ ^@ ^@ ^@ ^@ ^@ ^@ ^@ ^@ ^@1^@
^@
^@

Reply:
Good finding. Encoding. Now I remember that we had a similar issue when we was developing Raft

Reply:
@elirangwong

Please test again with this fix.

Thank you

Reply:
@crramirez Hello, I tried the latest fix.  It is still not working, output is here:

++ which wsl.exe
+++ which wsl.exe
++ /mnt/c/WINDOWS/system32/wsl.exe -l -v
+++ wsl.exe -l -v
+++ sed -e 's|\r||g' -e 's|\x00||g'
+++ grep WLinux
++++ locale charmap
+++ awk '{print $3}'
+++ iconv -f Unicode -t UTF-8
++ '[' Running -eq 2 ']'
-bash: [: Running: integer expression expected
++ export DISPLAY=:0
++ DISPLAY=:0

Reply:
@crramirez 

> wsl.exe -l -v > test.txt

> file -i text.txt
output:
> test.txt: application/octet-stream; charset=binary

> enca test.txt
output:
> Universal character set 2 bytes; UCS-2; BMP
>  CRLF line terminators
>  Byte order reversed in pairs (1,2 -> 2,1)

Reply:
@crramirez 

okay, after I tested your line piece by piece, I think I found the issue of the latest fix.

> wsl.exe -l -v

my output is:
  NAME      STATE           VERSION
* WLinux    Running         2
  WLE       Stopped         1

please note that I have "*" before WLinux, that's why "print $3" doesn't work for me
but "print $4" works in my case.

so $3 or $4 depends on if "*" is at the beginning ...

Reply:
simply change from $3 to $NF will work, I think, so part of your line should be:

$(wsl.exe -l -v 2>&1 | iconv -f Unicode -t $(locale charmap) | sed -e "s|\r||g" -e "s|\x00||g" | grep ${WSL_DISTRO_NAME} | awk '{print $NF}')

Reply:
OK, sounds logic. I switched to a simpler approach that will work for now, and if it changes in the future we can find a better solution. Check my latests commit

Reply:
@crramirez I think the latest fix haven't addressed the issue, raised by @hanubeki

I can see from his screenshot, there is no English words "Default Gateway" for your script to grep?

if I change my Windows settings to Chinese as deafult locale.  I think your script won't work, unless I manually put translation of "Default Gateway"?

Reply:
No problem he said here: https://github.com/WhitewaterFoundry/pengwin-base/pull/61#discussion_r354802502 that when the output is pipped it becomes English. Also added a fix for the IPv default gateway and in your case multiple addresses with Default Gateway

Reply:
@crramirez I think there is still an old issue with the latest 00-pengwin.sh.

After I use the latest 00-pengwin.sh, I cannot run anymore qt apps with QtWebEngine, e.g. falkon, unless I "manually" run export -n LIBGL_ALWAYS_INDIRECT.

remember the error "WebEngineContext used before QtWebEngine::initialize() or OpenGL context creation failed."?

in the latest 00-pengwin.sh, "export LIBGL_ALWAYS_INDIRECT=1" makes running those apps not possible even in WSL2.

Reply:
For LIBGL_ALWAYS_INDIRECT issue, some X server (such as X410) doesn't support OpenGL, so I had to unset LIBGL_ALWAYS_INDIRECT to some app that uses OpenGL.

Reply:
@hanubeki thank you.

after changing the line "export LIBGL_ALWAYS_INDIRECT=1" to "unset LIBGL_ALWAYS_INDIRECT", those apps work again.

thanks.